File: easy-2005-2/rem-periods.pl
#!/usr/bin/perl -w

use strict;
use warnings;

my $string = shift(@ARGV);

print rem_periods($string), "\n";

sub rem_periods
{
    my $string = shift;

    my $recurse;

    my @chars = split(//, $string);

    $recurse = sub {
        my ($arg) = (@_);
        my ($rest_of_chars) = [ @$arg];
        if (@$rest_of_chars == 0)
        {
            return ("", 0);
        }
        my $head = shift(@$rest_of_chars);
        my $tail = $rest_of_chars;
        my ($processed_string, $was_period_found) = $recurse->($tail);
        if ($was_period_found)
        {
            return ((($head eq "." ? "" : $head) . $processed_string), 1);
        }
        else
        {
            return ($head . $processed_string, ($head eq "."));
        }
    };

    return +($recurse->([@chars]))[0];
}

File: easy-2005-2/rem-periods.hs
module Main where

import System

main :: IO ()

main = do args <- getArgs
          putStr (rem_periods (head args))
          putStr "\n"

rem_periods :: String -> String
rem_periods mystring = (fst (recurse mystring)) where
    recurse :: String -> (String,Bool)
    recurse [] = ([],False)
    recurse (a:as) = (if was_period_found
                      then ((ret_string a), True) 
                      else ((a:processed_string), (a == '.'))
                      ) where
        (processed_string,was_period_found) = (recurse as)
        ret_string :: Char -> String
        ret_string '.' = processed_string
        ret_string _ = a:processed_string


